File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/onboarding/OnboardingScreen.kt
package lk.mageride.passenger.onboarding

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.passenger.R
import lk.mageride.passenger.shell.findActivity
import lk.mageride.passenger.ui.component.MageRideCta
import lk.mageride.passenger.ui.component.MageRideTextLink
import lk.mageride.passenger.ui.component.PagerDots
import lk.mageride.passenger.ui.component.SectionLabel
import lk.mageride.passenger.ui.component.SelectionBox
import lk.mageride.passenger.ui.theme.MageRideTheme

/**
 * SCR-PA-002 — the three-slide carousel and the language picker.
 *
 * **Get Started is pinned to the bottom, below the language list** — the prompt's fence, US-1.3's
 * own wording and the wireframe all say so, and a `Spacer(weight = 1f)` above it is what makes
 * that true at every supported height rather than only on the one the wireframe was drawn at.
 * D2' §SCR-PA-002's ASCII sketch still draws the CTA *above* the picker and the picker as a
 * `SegmentedButton`; the wireframe supersedes both. Recorded in the C077 handoff.
 *
 * **Skip and Get Started are the same action.** Neither skips the language: Sinhala is already the
 * highlighted default (AL-26) and the view model has already stored whatever is selected, so both
 * doors finish onboarding. The carousel is presentation only (BR-25.1) and gates nothing.
 *
 * @param onContinue Move to SCR-PA-003. The shell replaces the back stack — onboarding is a
 *   one-way door.
 */
@Composable
internal fun OnboardingScreen(onContinue: () -> Unit, model: OnboardingViewModel) {
    val state by model.state.collectAsStateWithLifecycle()
    val context = LocalContext.current
    val slides = rememberSlides(state)
    val pager = rememberPagerState(pageCount = { slides.size })

    val finish: () -> Unit = {
        // A language change only reaches `Resources` through `MainActivity.attachBaseContext`, so
        // the Activity is rebuilt before the next screen is drawn — otherwise a passenger who
        // picked සිංහල meets an English login screen. `recreate()` re-enters the graph at the
        // splash, which routes straight back here and then on, now in the chosen language.
        if (model.finish()) {
            // No Activity means a `@Preview`; continuing is the only sensible thing left.
            context.findActivity()?.recreate() ?: onContinue()
        } else {
            onContinue()
        }
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(MageRideTheme.spacing.md),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
    ) {
        Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
            MageRideTextLink(label = stringResource(R.string.onboarding_skip), onClick = finish)
        }

        HorizontalPager(state = pager, modifier = Modifier.fillMaxWidth()) { page ->
            SlidePage(slides[page])
        }

        PagerDots(
            count = slides.size,
            current = pager.currentPage,
            modifier = Modifier.fillMaxWidth().padding(vertical = MageRideTheme.spacing.xxs),
        )

        SectionLabel(text = stringResource(R.string.onboarding_language_label), centred = true)

        Column(verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs)) {
            // US-1.3's order — Sinhala, Tamil, English — is `LanguageChoices`, not the enum's.
            LanguageChoices.forEach { language ->
                SelectionBox(
                    label = language.endonym,
                    secondary = language.englishName,
                    selected = state.language == language,
                    onSelect = { model.select(language) },
                )
            }
        }

        // The pin. Everything above scrolls into whatever height is left; the CTA does not move.
        Box(modifier = Modifier.weight(1f))

        MageRideCta(label = stringResource(R.string.onboarding_get_started), onClick = finish)
    }
}

/** One page of the carousel. */
@Composable
private fun SlidePage(slide: FeatureSlide) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        lk.mageride.passenger.ui.component.IllustrationPanel(icon = slide.icon, caption = slide.caption)
        Text(
            text = slide.title,
            style = MaterialTheme.typography.headlineMedium,
            color = MaterialTheme.colorScheme.onSurface,
            textAlign = TextAlign.Center,
        )
        Text(
            text = slide.body,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
        )
    }
}

/**
 * content-svc's slides in the chosen language, or the bundled ones.
 *
 * Resolved here rather than in the view model because the fallback's copy is a **string
 * resource** — `stringResource` is composable, and a `@StringRes` int that had to be resolved in a
 * view model would need a `Context` there. Switching language re-runs this, which is what makes
 * the carousel change with the picker without a second fetch.
 */
@Composable
private fun rememberSlides(state: OnboardingState): List<FeatureSlide> = if (state.slides.isEmpty()) {
    FeatureSlides.Fallback.map {
        FeatureSlide(
            title = stringResource(it.title),
            body = stringResource(it.body),
            caption = stringResource(it.caption),
            icon = it.icon,
        )
    }
} else {
    state.slides.map { FeatureSlides.resolve(it, state.language) }
}

File: apps/passenger-ios/PassengerApp/Onboarding/OnboardingScreen.swift
import MageRideShared
import SwiftUI

/// **SCR-PI-002 · onboarding + language** — US-1.2's three-slide tutorial and US-1.3's picker.
///
/// The wireframe's column, top to bottom: a `Skip` action in the nav row, the illustration panel,
/// the headline and its sentence, the dots, a `Language` label, three selectable rows, **a spacer**,
/// and `Get Started`.
///
/// **Get Started is pinned to the bottom, below the language rows** — the cell's own `Δ iOS` clause
/// says so in as many words (*"pinned to the bottom of the screen (safe-area inset, below the
/// language rows after the spacer)"*) and the component prompt makes it a fence. The `Spacer()` above
/// it is what makes "pinned" true at every supported height rather than only at the one the
/// wireframe was drawn at; C077 made the same call with a weighted `Box`.
///
/// **D2' §SCR-PA-002 disagrees with the wireframe on two counts and the wireframe wins**: its ASCII
/// sketch draws the CTA *above* the language row, and its component table names a `SegmentedButton`
/// where US-1.3 asks for *"vertical selectable boxes, one per row"*. The wireframes are the
/// team-approved baseline; C077 recorded the same conflict and it still needs a micro-change-set.
///
/// **`TabView(.page)` is the carousel** — the cell's own clause again — with the system's own index
/// dots turned off, because the wireframe draws an 18pt `primary` pill for the active one and the
/// system draws a circle in a colour a `TabView` decides.
@MainActor
struct OnboardingScreen: View {

    @StateObject private var model: OnboardingModel

    private let onContinue: () -> Void

    init(repository: OnboardingRepository, onContinue: @escaping () -> Void) {
        _model = StateObject(wrappedValue: OnboardingModel(onboarding: repository))
        self.onContinue = onContinue
    }

    var body: some View {
        VStack(spacing: MageRideSpacing.sm) {
            skip

            carousel

            PageDots(count: model.state.slides.count, current: model.state.page)

            SectionLabel(key: "onboarding_language_label", alignment: .center)
                .padding(.top, MageRideSpacing.xxs)

            languages

            // The fence. Everything above sits at the top of the screen; Get Started sits at the
            // bottom whatever the handset's height.
            Spacer(minLength: MageRideSpacing.md)

            Button(action: proceed) {
                Text(key: "onboarding_get_started")
            }
            .buttonStyle(.mageCta)
        }
        .padding(.horizontal, MageRideSpacing.md)
        .padding(.bottom, MageRideSpacing.md)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(MageRideColor.surface)
        .task { await model.start() }
    }

    /// The wireframe's `.navtop` — a spacer and a trailing `Skip`.
    ///
    /// Skip and Get Started are the **same action**: neither skips the language, because there is
    /// nothing to skip (AL-26 pre-selects Sinhala and every tap has already been stored). See
    /// ``OnboardingModel/finish()``.
    private var skip: some View {
        HStack {
            Spacer()
            TextLink(key: "onboarding_skip", action: proceed)
        }
    }

    /// **The pages are walked by index rather than by `enumerated()`** (Δ C101).
    ///
    /// This loop was `ForEach(Array(model.state.slides.enumerated()), id: \.element.id)`, and
    /// `\.element` is a key path into a **tuple** — which the C087 finding this repository carries
    /// says does not compile, and which `ModeFilterSheet`, `PaymentMethodScreen` and
    /// `HistoryControls` all deliberately avoid for that reason. No host has built this target, so
    /// nothing has settled which reading is right; walking indices is correct under either, which is
    /// why it is written this way rather than left for a macOS runner to adjudicate. (`apps/driver-ios`
    /// has six more of the same shape — that target's to settle, not this one's.)
    ///
    /// The index cannot simply be dropped: `.tag(index)` is what `pageBinding` selects on, so a page
    /// is identified by its position rather than by its slide.
    private var carousel: some View {
        TabView(selection: pageBinding) {
            ForEach(model.state.slides.indices, id: \.self) { index in
                slide(model.state.slides[index])
                    .tag(index)
                    // One announcement per slide: VoiceOver reads the headline and its sentence
                    // together rather than as two elements a reader has to assemble (US-19.1/19.2).
                    .accessibilityElement(children: .combine)
            }
        }
        .tabViewStyle(.page(indexDisplayMode: .never))
        .frame(height: carouselHeight)
    }

    private func slide(_ slide: FeatureSlide) -> some View {
        VStack(spacing: MageRideSpacing.sm) {
            IllustrationPanel(
                symbolName: slide.symbolName,
                caption: slide.caption,
                height: MageRideControl.illustrationPanel
            )

            Text(slide.title)
                .mageFont(.headline)
                .foregroundStyle(MageRideColor.onSurface)
                .multilineTextAlignment(.center)

            Text(slide.body)
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
                .multilineTextAlignment(.center)

            Spacer(minLength: 0)
        }
    }

    /// US-1.3's rows, in the order the story fixes: Sinhala, Tamil, English.
    private var languages: some View {
        VStack(spacing: MageRideSpacing.xs) {
            ForEach(LanguageDisplay.choices, id: \.self) { language in
                SelectionRow(
                    label: LanguageDisplay.endonym(language),
                    secondary: LanguageDisplay.englishName(language),
                    isSelected: language == model.state.language
                ) {
                    // `.selection` on a language change — the cell's own clause. The haptic is the
                    // acknowledgement, because the row's own highlight moves at the same instant the
                    // whole screen re-renders in the new language and the eye has nothing to fix on.
                    UISelectionFeedbackGenerator().selectionChanged()
                    model.select(language)
                }
            }
        }
    }

    private var pageBinding: Binding<Int> {
        Binding(get: { model.state.page }, set: { model.onPageChanged($0) })
    }

    private func proceed() {
        model.finish()
        onContinue()
    }

    /// Room for the panel, a two-line headline and a two-line sentence at the largest Dynamic Type
    /// size the wireframe's layout survives.
    ///
    /// A `TabView` needs a bound height — it fills whatever it is given — so this is a measurement of
    /// one control rather than spacing between two, which is why it is not a ``MageRideSpacing``
    /// token. It is the only fixed height on this screen; everything else grows with the text.
    private let carouselHeight: CGFloat = 260
}
